File: leetcode3/ShortestImpossibleSequenceofRolls/a.go
/* Copyright (C) 2022 by iamslash */

package main

import "fmt"

// 157ms 78.43% 8.5MB 98.04%
// hash set
// O(N) O(K)
func shortestSequence(rolls []int, k int) int {
    var ans = 1
    var diceSet = make(map[int]struct{})
    for _, e := range rolls {
        diceSet[e] = struct{}{}
        if len(diceSet) == k {
            diceSet = make(map[int]struct{})
            ans++
        }
    }
    return ans
}

func main() {
    fmt.Println("hello world")
}

File: leetcode3/ShortestImpossibleSequenceofRolls/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2022 by iamslash

from typing import List

# 2251ms 6.48% 28.3MB 38.25%
# hash set
# O(N) O(K)
class Solution:
    def shortestSequence(self, rolls: List[int], k: int) -> int:
        ans = 1
        diceSet = set()
        for roll in rolls:
            diceSet.add(roll)
            if len(diceSet) == k:
                diceSet = set()
                ans += 1
        return ans
